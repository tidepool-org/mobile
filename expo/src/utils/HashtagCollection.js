import twitter from "twitter-text";

class HashtagCollection {
  static defaultHashtags = [
    "#exercise",
    "#meal",
    "#sitechange",
    "#sensorchange",
    "#juicebox",
    "#devicesetting",
  ];

  hashtagsWithCountSortedByCount = [];
  hashtagsWithCountMap = new Map();

  constructor() {
    this.resetToDefaultHashtags();
  }

  resetToDefaultHashtags() {
    this.reset();
    this.addDefaultHashtagsToMap();
    this.updateSortedHashtags();
  }

  reloadHashtagsForObjectsWithText({
    arrayOfObjectsWithText,
    textPropertyName,
  }) {
    this.reset();
    if (arrayOfObjectsWithText && arrayOfObjectsWithText.length > 0) {
      arrayOfObjectsWithText.forEach(objectWithText => {
        this.addHashtagsForTextToMap(objectWithText[textPropertyName]);
      });
    }
    this.addDefaultHashtagsToMap();
    this.updateSortedHashtags();
  }

  updateHashtagsForText({
    oldObjectWithText,
    newObjectWithText,
    textPropertyName,
  }) {
    if (newObjectWithText) {
      this.addHashtagsForTextToMap(newObjectWithText[textPropertyName]);
    }
    if (oldObjectWithText) {
      this.removeHashtagsForTextFromMap(oldObjectWithText[textPropertyName]);
    }
    this.updateSortedHashtags();
  }

  // Private

  reset() {
    this.hashtagsWithCountSortedByCount = [];
    this.hashtagsWithCountMap = new Map();
  }

  updateSortedHashtags() {
    this.hashtagsWithCountSortedByCount = Array.from(this.hashtagsWithCountMap);
    this.hashtagsWithCountSortedByCount.sort(
      (hashtagWithCount1, hashtagWithCount2) => {
        const count1 = hashtagWithCount1[1];
        const count2 = hashtagWithCount2[1];
        if (count2 > count1) {
          return 1;
        } else if (count2 < count1) {
          return -1;
        }
        return 0;
      }
    );
  }

  addDefaultHashtagsToMap() {
    HashtagCollection.defaultHashtags.forEach(hashtag => {
      this.addHashtagToMap(hashtag);
    });
  }

  addHashtagsForTextToMap(text) {
    const extractedHashtags = twitter.extractHashtags(text);
    extractedHashtags.forEach(hashtag => {
      this.addHashtagToMap(`#${hashtag}`);
    });
  }

  removeHashtagsForTextFromMap(text) {
    const extractedHashtags = twitter.extractHashtags(text);
    extractedHashtags.forEach(hashtag => {
      this.removeHashtagFromMap(`#${hashtag}`);
    });
  }

  addHashtagToMap(hashtag) {
    const count = this.hashtagsWithCountMap.get(hashtag) || 0;
    this.hashtagsWithCountMap.set(hashtag, count + 1);
  }

  removeHashtagFromMap(hashtag) {
    const count = this.hashtagsWithCountMap.get(hashtag);
    if (count) {
      if (count > 1) {
        this.hashtagsWithCountMap.set(hashtag, count - 1);
      } else {
        this.hashtagsWithCountMap.delete(hashtag);
      }
    }
  }
}

export default HashtagCollection;
