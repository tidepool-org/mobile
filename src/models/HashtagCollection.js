import twitter from "twitter-text";

class HashtagCollection {
  static defaultHashtags = [
    "#exercise",
    "#meal",
    "#sitechange",
    "#sensorchange",
    "#juicebox",
    "#devicesetting",
  ];

  hashtagsSortedByCount = [];
  hashtagsWithCountSortedByCount = [];
  hashtagsWithCountMap = new Map();

  constructor() {
    this.resetToDefaultHashtags();
  }

  resetToDefaultHashtags() {
    this.reset();
    this.addDefaultHashtagsToMap();
    this.updateSortedHashtags();
  }

  reloadHashtagsForObjectsWithText({ objectsWithText, textPropertyName }) {
    this.reset();
    if (objectsWithText && objectsWithText.length > 0) {
      objectsWithText.forEach(objectWithText => {
        this.addHashtagsForTextToMap(objectWithText[textPropertyName]);
      });
    }
    this.addDefaultHashtagsToMap();
    this.updateSortedHashtags();
  }

  updateHashtagsForText({
    objectWithTextToRemove,
    objectWithTextToAdd,
    textPropertyName,
  }) {
    if (objectWithTextToAdd) {
      this.addHashtagsForTextToMap(objectWithTextToAdd[textPropertyName]);
    }
    if (objectWithTextToRemove) {
      this.removeHashtagsForTextFromMap(
        objectWithTextToRemove[textPropertyName]
      );
    }
    this.updateSortedHashtags();
  }

  // Private

  reset() {
    this.hashtagsSortedByCount = [];
    this.hashtagsWithCountSortedByCount = [];
    this.hashtagsWithCountMap = new Map();
  }

  updateSortedHashtags() {
    this.hashtagsWithCountSortedByCount = Array.from(this.hashtagsWithCountMap);
    this.hashtagsWithCountSortedByCount.sort(
      (hashtagWithCount1, hashtagWithCount2) => {
        const count1 = hashtagWithCount1[1];
        const count2 = hashtagWithCount2[1];
        if (count2 > count1) {
          return 1;
        } else if (count2 < count1) {
          return -1;
        }
        return 0;
      }
    );
    this.hashtagsSortedByCount = this.hashtagsWithCountSortedByCount.map(
      hashtagWithCount => hashtagWithCount[0]
    );
  }

  addDefaultHashtagsToMap() {
    HashtagCollection.defaultHashtags.forEach(hashtag => {
      this.addHashtagToMap(hashtag);
    });
  }

  addHashtagsForTextToMap(text) {
    const extractedHashtags = twitter.extractHashtags(text);
    extractedHashtags.forEach(hashtag => {
      this.addHashtagToMap(`#${hashtag}`);
    });
  }

  removeHashtagsForTextFromMap(text) {
    const extractedHashtags = twitter.extractHashtags(text);
    extractedHashtags.forEach(hashtag => {
      this.removeHashtagFromMap(`#${hashtag}`);
    });
  }

  addHashtagToMap(hashtag) {
    const count = this.hashtagsWithCountMap.get(hashtag) || 0;
    this.hashtagsWithCountMap.set(hashtag, count + 1);
  }

  removeHashtagFromMap(hashtag) {
    const count = this.hashtagsWithCountMap.get(hashtag);
    if (count) {
      if (count > 1) {
        this.hashtagsWithCountMap.set(hashtag, count - 1);
      } else {
        this.hashtagsWithCountMap.delete(hashtag);
      }
    }
  }
}

export default HashtagCollection;
