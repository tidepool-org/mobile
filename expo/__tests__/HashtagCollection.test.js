/* eslint-disable no-undef */
/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable no-console */

import HashtagCollection from "../src/utils/HashtagCollection";

it("has default hashtags for new collection", () => {
  const hashtagCollection = new HashtagCollection();
  const {
    hashtagsWithCountSortedByCount,
    hashtagsWithCountMap,
  } = hashtagCollection;

  expect(hashtagsWithCountSortedByCount.length).toBeGreaterThan(0);
  expect(hashtagsWithCountMap.size).toBeGreaterThan(0);
  expect(hashtagsWithCountSortedByCount.length).toEqual(
    hashtagsWithCountMap.size
  );
});

it("has same snapshot default hashtags for new collection", () => {
  const hashtagCollection = new HashtagCollection();
  const {
    hashtagsWithCountSortedByCount,
    hashtagsWithCountMap,
  } = hashtagCollection;

  expect(hashtagsWithCountSortedByCount).toMatchSnapshot();
  expect(hashtagsWithCountMap).toMatchSnapshot();
});

it("sorts hashtags by count", () => {
  const hashtagCollection = new HashtagCollection();
  hashtagCollection.updateHashtagsForText({
    newObjectWithText: { text: "#hash2 #hash2 #hash1 #hash1 #hash1 #hash3" },
    textPropertyName: "text",
  });

  const { hashtagsWithCountSortedByCount } = hashtagCollection;
  expect(hashtagsWithCountSortedByCount[0][0]).toEqual("#hash1");
  expect(hashtagsWithCountSortedByCount[1][0]).toEqual("#hash2");
});

it("deletes hashtags", () => {
  const hashtagCollection = new HashtagCollection();
  hashtagCollection.updateHashtagsForText({
    newObjectWithText: { text: "#hash2 #hash2 #hash1 #hash1 #hash1 #hash3" },
    textPropertyName: "text",
  });
  hashtagCollection.updateHashtagsForText({
    oldObjectWithText: { text: "#hash2 #hash2 #hash1 #hash1 #hash1 #hash3" },
    textPropertyName: "text",
  });
  hashtagCollection.updateHashtagsForText({
    newObjectWithText: { text: "#hash2 #hash2 #hash1 #hash1 #hash1 #hash3" },
    textPropertyName: "text",
  });

  const { hashtagsWithCountSortedByCount } = hashtagCollection;
  expect(hashtagsWithCountSortedByCount[0][0]).toEqual("#hash1");
  expect(hashtagsWithCountSortedByCount[0][1]).toEqual(3);
});
