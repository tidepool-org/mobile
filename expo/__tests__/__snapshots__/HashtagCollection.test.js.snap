// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has same snapshot default hashtags for new collection 1`] = `
Array [
  Array [
    "#exercise",
    1,
  ],
  Array [
    "#meal",
    1,
  ],
  Array [
    "#sitechange",
    1,
  ],
  Array [
    "#sensorchange",
    1,
  ],
  Array [
    "#juicebox",
    1,
  ],
  Array [
    "#devicesetting",
    1,
  ],
]
`;

exports[`has same snapshot default hashtags for new collection 2`] = `
Map {
  "#exercise" => 1,
  "#meal" => 1,
  "#sitechange" => 1,
  "#sensorchange" => 1,
  "#juicebox" => 1,
  "#devicesetting" => 1,
}
`;
